flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/PendingCheckpoint.java

```diff
@@ -308,6 +316,25 @@ public CompletedCheckpoint finalizeCheckpoint(
 
             // make sure we fulfill the promise with an exception if something fails
             try {
+                // Completes the operator state for the fully finished operators
+                for (ExecutionJobVertex jobVertex : checkpointBrief.getFullyFinishedJobVertex()) {
+                    for (OperatorIDPair operatorID : jobVertex.getOperatorIDs()) {
+                        OperatorState operatorState =
+                                operatorStates.get(operatorID.getGeneratedOperatorID());
+
+                        if (operatorState == null) {
+                            operatorState =
+                                    new OperatorState(
+                                            operatorID.getGeneratedOperatorID(),
+                                            jobVertex.getParallelism(),
+                                            jobVertex.getMaxParallelism());
+                            operatorStates.put(operatorID.getGeneratedOperatorID(), operatorState);
+                        }
+
+                        operatorState.markedFullyFinished();
```

Comment:
IIUC, the actual state should not be present for finished operators for this checkpoint, right?
Can we add a check for this?

Otherwise, we need to discard previous states.

Reply:
It should be always null, I add the check for it.

Reply:
I think the check is still missing.

Reply:
Very sorry for miss the check when migrating the code... I re-added the check.